### PR TITLE
[InstCombine] Fold expression using basic properties of floor and ceiling function

### PR DESCRIPTION
**Pull request by c8ef**
alive2: ~~https://alive2.llvm.org/ce/z/Ag3Ki7~~ https://alive2.llvm.org/ce/z/ywP5t2
related: #76438

This patch adds the following foldings: `floor(x) <= x --> true` and `x <= ceil(x) --> true`. We leverage the properties of these math functions and ensure there is no floating point input of `nan`.

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4134]: These queries can be expensive and you should try to defer them as late as possible (and prefer to early exit on the RHS query before the LHS, as the RHS is canonically cheaper)

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4143]: Separate lines for each of these comments

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4145]: Match the intrinsic before the expensive class query

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4159]: You wouldn't need the known never nan checks if you can replace with an fcmp ord/uno, but then you have to create a new instruction. Maybe it would be better to just handle this in instcombine?

File: llvm/test/Transforms/InstSimplify/fp-floor-ceil.ll
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/fp-floor-ceil.ll:8]: Use named values in tests.

Also test a case where the known-never-nan comes from the source value, not the instruction flag (e.g. nofpclass(nan) on the incoming function argument)

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4159]: You don't need the floor at all: https://alive2.llvm.org/ce/z/E4erNR

**Comment by dtcxzyw** [comment]
> Hi @dtcxzyw, could you help me verify if the transformation in the patch has any real-world use cases?

This pattern is simple so I think it is ok to handle this regardless of its real-world usefulness.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8191]: Do the pred checks first?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8193]: You do not need to call isKnownNeverNaN. The only reason it was useful in instsimplify was you could not insert the new fcmp required to preserve the nan behavior

File: llvm/test/Transforms/InstCombine/fp-floor-ceil.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fp-floor-ceil.ll:194]: Needs tests showing fast math flag preservation on the fcmp

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8193]: It needs some adjustment, but you can avoid it: https://alive2.llvm.org/ce/z/YZfsRb

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8191]: Yes

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Please update the alive2 link in your PR description.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8190]: I think it would be cleaner if you swapped operands so floor/ceil is always on the left, instead of always checking both variants.

File: llvm/test/Transforms/InstCombine/fp-floor-ceil.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fp-floor-ceil.ll:194]: Negative tests for the other compare types, like oeq/one/une/ord/uno

File: llvm/test/Transforms/InstCombine/fp-floor-ceil.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fp-floor-ceil.ll:106]: This is the only flag preserving test, but it only covers oge. Should have one for each of these handled cases (i.e. have a test with a flag with ole, oge, ule, ugt, uge, and ult)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8195]: Instead of pre-matching all the cases, you could match the preferred direction and then only try the other match if they both failed. You can also then directly assign floor_x / ceil_x from the match function when commuting.

Also variable names should be camel case

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8225]:
```suggestion
                          ConstantFP::getInfinity(OpType, true), RHS,
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8182]:
```suggestion
                                             InstCombinerImpl &IC) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8229]:
```suggestion
    // fcmp ugt floor(x), x => fcmp uno x, 0
```
https://alive2.llvm.org/ce/z/MpPEy7

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8228]: I mean you should modify both the comment and the transform below.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Please wait for additional approval from other reviewers :)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8196]: FWIW, I think it's usually more elegant to do this by calling `foldFCmpWithFloorAndCeil(Pred, Op0, Op1)` and then `foldFCmpWithFloorAndCeil(getSwappedPredicate(), Op1, Op0)`, so the method itself only has to deal with one variant. But the way you did it is fine as well.

**Review by nikic — COMMENTED** [review]
Looks fine to me, but I didn't review in detail.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8244]: This isn't handling the cases that use infinities?

**Review by arsenm — APPROVED** [review]

**Pull request merged**
